packages/eslint-plugin/src/rules/no-unnecessary-type-constraint.ts

```diff
@@ -0,0 +1,116 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import * as semver from 'semver';
+import * as ts from 'typescript';
+import * as util from '../util';
+
+type MakeRequired<Base, Key extends keyof Base> = Omit<Base, Key> &
+  Required<Pick<Base, Key>>;
+
+type TypeParameterWithConstraint = MakeRequired<
+  TSESTree.TSTypeParameter,
+  'constraint'
+>;
+
+type KeywordFilter = (type: ts.Type) => boolean;
+
+const is3dot5 = semver.satisfies(
+  ts.version,
+  `>= 3.5.0 || >= 3.5.1-rc || >= 3.5.0-beta`,
+  {
+    includePrerelease: true,
+  },
+);
+
+const is3dot9 =
+  is3dot5 &&
+  semver.satisfies(ts.version, `>= 3.9.0 || >= 3.9.1-rc || >= 3.9.0-beta`, {
+    includePrerelease: true,
+  });
+
+export default util.createRule({
+  name: 'no-unnecessary-type-constraint',
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Disallows unnecessary constraints on generic types',
+      recommended: false,
+      requiresTypeChecking: true,
+      suggestion: true,
+    },
+    fixable: 'code',
+    messages: {
+      unnecessaryConstraint:
+        'Constraining a generic type to {{constraint}} does nothing and is unnecessary.',
```

Comment:
might be good to add the generic name to the message.
Just makes it clearer for CLI runs which provide no other context apart from line numbers.

Reply:
Sorry, what do you mean by "generic name" here?

Reply:
For `T extends any` the message could be
``Constraining the generic type `T` to `any` does nothing and is unnecessary.``